src/circular_buffer.jl

```diff
@@ -193,3 +193,46 @@ capacity(cb::CircularBuffer) = cb.capacity
 Test whether the buffer is full.
 """
 isfull(cb::CircularBuffer) = length(cb) == cb.capacity
+
+"""
+    first(cb)
+
+Get the first element of CircularBuffer.
+"""
+Base.@propagate_inbounds function Base.first(cb::CircularBuffer)
+    @boundscheck if cb.length == 0
+        throw(BoundsError(cb, 1))
+    end
+    cb.buffer[cb.first]
+end
+"""
+    last(cb)
+
+Get the last element of CircularBuffer.
+"""
+Base.@propagate_inbounds function Base.last(cb::CircularBuffer)
+    @boundscheck if cb.length == 0
+        throw(BoundsError(cb, 1))
+    end
+    cb.buffer[_buffer_index(cb, cb.length)]
+end
+"""
+    resize!(cb, n)
+
+Resize CircularBuffer to the maximum capacity of n elements. If n is smaller than the current buffer length, the first n elements will be retained.
+"""
+function Base.resize!(cb::CircularBuffer{T}, n::Int) where {T}
+    if n != capacity(cb)
+        buf_new = Vector{T}(undef, n)
+        len_new = min(length(cb), n)
+        for i = 1 : len_new
+            @inbounds buf_new[i] = cb[i]
+        end
```

Comment:
Why not just call `resize!(cb.buffer)`

Reply:
Because `cb.first` can point to any index within old size.

Reply:
Actually, it can be `circshift`'ed: [first:length, 1:first]

Reply:
You can use `splice!` to delete inplace,
It should even avoid copies when only deleting from one end or the other

Reply:
In the most complex case, where buffer is not full, 
e.g. `cb.first` = 5, `cb.length` = 6, `cb.capacity` = 8: [5 6 x x 1 2 3 4], and:
a) resized to size-1 => [5 6 x 1 2 3 4] (deleted from middle)
b) resized to size-5 => [1 2 3] (deleted from both sides)
c) resized to size+1 => [1 2 3 4 5 6 x x x] (circshifted and added to the end)
So, should we really detect all possible cases, or end-up just with copying to another array?

Reply:
More broadly it is unclear,
should `resize!` upwards only increase the capacity
and not the length  (docstring currently says that, but code increases capacity and length).

It is not clear what this is for to me, and if this is the best  way to do this,
and wether it belongs here, or in the `CircularDeque`.

Might be best to let this PR just do `first` and `last`
and make a follow up to add this.

Reply:
Ok, I've reverted that change, will be another PR.